src/System.Private.ServiceModel/tests/Stress/SharedPoolsOfWCFObjects/PoolOfThings.cs

```diff
@@ -0,0 +1,374 @@
+                {
+                    try
+                    {
+                        //Interlocked.Increment(ref _stats.FactoriesClosed);
+                        await destroyInstanceAsync(instance);
+                    }
+                    catch (Exception e)
+                    {
+                        //Interlocked.Increment(ref _stats.ExceptionsClosingFactories);
+                    }
+                }
+            }
+        }
+        #endregion
+    }
+
+    /// <summary>
+    /// This helper classes below group a factory F with a pool of objects O created with the help of the factory
+    /// Keeping them together makes it easier to create and cleanup pools of pools of objects
+    /// Unlike the traditional factory patters, rather than demanding the factory F to be able to create/destroy 
+    /// instances of O we take delegates to create/destroy instances of O as constructor parameters.
+    /// This allows us to decouple F and O and makes it easy to introduce additional wrappers around the instances F creates
+    /// </summary>
+    public class FactoryAndPoolOfItsObjects<F, O>
+        where F : class
+        where O : class
+    {
+        private Action<F> _destroyFactoryInstance;
+        public F Factory { get; set; }
+        public PoolOfThings<O> ObjectsPool { get; set; }
+
+        /// <summary>
+        /// The constructor simply creates an empty pool of objects and stores delegates to create/destroy instances
+        /// </summary>
+        /// <param name="createFactoryInstance"> A func to create an instance of F. </param>
+        /// <param name="destroyFactoryInstance"> An action to destroy an instance of F</param>
+        /// <param name="maxPooledObjects"> Max size of the pool of objects O </param>
+        /// <param name="createObject"> A func to create all instances of O </param>
+        /// <param name="destroyObject"> An action to destroy all instances of O</param>
+        public FactoryAndPoolOfItsObjects(Func<F> createFactoryInstance, Action<F> destroyFactoryInstance, int maxPooledObjects, Func<F, O> createObject, Action<O> destroyObject)
+        {
+            _destroyFactoryInstance = destroyFactoryInstance;
+            Factory = createFactoryInstance();
+            ObjectsPool = new PoolOfThings<O>(
+                maxSize: maxPooledObjects,
+                createInstance: () =>
+                {
+                    // PoolOfThings lets us use the instances while destroying them
+                    // Destroy may null Factory out so we save a local copy here
+                    var f = Factory;
+                    return f != null ? createObject(f) : null;
+                },
+                destroyInstance: destroyObject);
+        }
+
+        public void Destroy()
+        {
+            // should never be recycled more than once
+            if (Factory == null)
+            {
+                System.Diagnostics.Debugger.Break();
+            }
```

Comment:
For future consideration -- breaking to the debugger should probably be avoided in stress runs unless you are debugging some problem.  It could stop the test run for what is arguably not a huge issue (double destroy).